xdsl/transforms/convert_memref_to_ptr.py

```diff
@@ -153,12 +154,202 @@ def match_and_rewrite(self, op: memref.LoadOp, rewriter: PatternRewriter, /):
+
+        rewriter.replace_matched_op(func.ReturnOp(*new_arguments))
+
+
+@dataclass
+class LowerMemrefFuncCallPattern(RewritePattern):
+    @op_type_rewrite_pattern
+    def match_and_rewrite(self, op: func.CallOp, rewriter: PatternRewriter, /):
+        if not any(
+            isinstance(arg.type, memref.MemRefType) for arg in op.arguments
+        ) and not any(isinstance(type, memref.MemRefType) for type in op.result_types):
+            return
+
+        # rewrite arguments
+        insert_point = InsertPoint.before(op)
+        new_arguments: list[SSAValue] = []
+
+        # insert `memref -> ptr` casts for memref arguments values
+        for argument in op.arguments:
+            if isinstance(argument.type, memref.MemRefType):
+                rewriter.insert_op(
+                    cast_op := builtin.UnrealizedConversionCastOp.get(
+                        [argument], [ptr.PtrType()]
+                    ),
+                    insert_point,
+                )
+                new_arguments.append(cast_op.results[0])
+            else:
+                new_arguments.append(argument)
+
+        insert_point = InsertPoint.after(op)
+        new_results: list[SSAValue] = []
+
+        #  insert `ptr -> memref` casts for return values
+        for result in op.results:
+            if isinstance(result.type, memref.MemRefType):
+                rewriter.insert_op(
+                    cast_op := builtin.UnrealizedConversionCastOp.get(
+                        [result],
+                        # TODO: annoying pyright warnings - Sasha, pls help
+                        [result.type],  # pyright: ignore[reportUnknownMemberType,reportUnknownArgumentType]
+                    ),
+                    insert_point,
+                )
+                new_results.append(cast_op.results[0])
+            else:
+                new_results.append(result)
+
+        new_return_types = [
+            ptr.PtrType() if isinstance(type, memref.MemRefType) else type
+            for type in op.result_types
+        ]
+
+        rewriter.replace_matched_op(
+            func.CallOp(op.callee, new_arguments, new_return_types)
+        )
+
+
+class ReconcileUnrealizedPtrCasts(RewritePattern):
+    """
+    Eliminates two variants of unrealized ptr casts:
+    - `ptr_xdsl.ptr -> memref.MemRef -> ptr_xdsl.ptr`;
+    - `ptr_xdsl.ptr -> memref.memref` where all uses are `ToPtrOp` operations.
+    """
+
+    @op_type_rewrite_pattern
+    def match_and_rewrite(
+        self, op: builtin.UnrealizedConversionCastOp, rewriter: PatternRewriter, /
+    ):
+        # preconditions
+        if (
+            len(op.inputs) != 1
+            or len(op.outputs) != 1
+            or not isinstance(op.inputs[0].type, ptr.PtrType)
+        ):
+            return
+
+        if not isinstance(op.outputs[0].type, memref.MemRefType):
+            return
+
+        # erase ptr -> memref -> ptr cast pairs
+        uses = [use for use in op.outputs[0].uses]
```

Comment:
One more thing is that I recommend tuples over lists as they're more efficient and play better with the type system